openfisca_uk/calibration/reweighting.py

```diff
@@ -0,0 +1,358 @@
+                if i < num_thresholds - 1
+                else np.inf
+            )
+            income = sim.calc("total_income", period=year)
+            person_in_range = (income >= lower_threshold) & (
+                income < upper_threshold
+            )
+            income_tax_in_range = (
+                sim.calc("income_tax", period=year) * person_in_range
+            )
+            household_income_tax = sim.map_to(
+                income_tax_in_range, "person", "household"
+            )
+            aggregate = tf.reduce_sum(modified_weights * household_income_tax)
+            target = brackets[i].amount(instant_str)
+            l += (
+                AGGREGATE_ERROR_PENALTY
+                * target
+                * ((aggregate / target) - 1) ** 2
+                / num_thresholds
+            )
+
+        # Population by age
+
+        brackets = statistics.population_by_age.brackets
+        num_thresholds = len(brackets)
+        for i in range(num_thresholds):
+            lower_threshold = brackets[i].threshold(instant_str)
+            upper_threshold = (
+                brackets[i + 1].threshold(instant_str)
+                if i < num_thresholds - 1
+                else np.inf
+            )
+            age = sim.calc("age", period=year)
+            person_in_range = (age >= lower_threshold) & (
+                age < upper_threshold
+            )
+            household_count = sim.map_to(
+                person_in_range, "person", "household"
+            )
+            population = tf.reduce_sum(modified_weights * household_count)
+            if year > 2019:
+                current_population = sum(
+                    [
+                        getattr(statistic_set.populations, region)
+                        for region in regions
+                    ]
+                )
+                last_year_population = sum(
+                    [
+                        getattr(
+                            statistics(f"{year-1}-01-01").populations, region
+                        )
+                        for region in regions
+                    ]
+                )
+                population_increase_ratio = (
+                    current_population / last_year_population
+                )
+            else:
+                population_increase_ratio = 1
+            target = (
+                brackets[i].amount(instant_str) * population_increase_ratio
+            )
+            l += (
+                POPULATION_ERROR_PENALTY
+                * AGE_DISTRIBUTION_PENALTY
+                * target
+                * ((population / target) - 1) ** 2
+                / num_thresholds
+            )
+
+        l += (
+            POPULATION_ERROR_PENALTY
+            * PARTICIPATION_ERROR_PENALTY
+            * weights.sum()
+            * ((tf.reduce_sum(modified_weights) / weights.sum()) - 1) ** 2
+        )
+
+    # Add penalty for weight changes
+    if include_modification_penalty:
```

Comment:
Why the `if` statement instead of setting the penalty to zero?

Reply:
Code-character-count efficiency? No real preference